sdk/spring/azure-spring-boot/src/main/java/com/azure/spring/autoconfigure/aad/AccessTokenManager.java

```diff
@@ -0,0 +1,230 @@
+// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+package com.azure.spring.autoconfigure.aad;
+
+import com.microsoft.aad.msal4j.ClientCredentialFactory;
+import com.microsoft.aad.msal4j.ConfidentialClientApplication;
+import com.microsoft.aad.msal4j.IAuthenticationResult;
+import com.microsoft.aad.msal4j.IClientCredential;
+import com.microsoft.aad.msal4j.MsalServiceException;
+import com.microsoft.aad.msal4j.OnBehalfOfParameters;
+import com.microsoft.aad.msal4j.UserAssertion;
+import org.slf4j.Logger;
+import org.slf4j.LoggerFactory;
+import org.springframework.security.core.Authentication;
+import org.springframework.security.core.context.SecurityContext;
+import org.springframework.security.core.context.SecurityContextHolder;
+import org.springframework.security.oauth2.core.AbstractOAuth2Token;
+import org.springframework.security.oauth2.core.oidc.user.OidcUser;
+import org.springframework.web.context.request.RequestContextHolder;
+import org.springframework.web.context.request.ServletRequestAttributes;
+
+import javax.naming.ServiceUnavailableException;
+import javax.servlet.http.HttpServletRequest;
+import java.net.MalformedURLException;
+import java.util.Arrays;
+import java.util.Date;
+import java.util.HashSet;
+import java.util.Locale;
+import java.util.Optional;
+import java.util.Set;
+import java.util.UUID;
+import java.util.concurrent.ExecutionException;
+import java.util.stream.Collectors;
+
+import static com.azure.spring.autoconfigure.aad.Scopes.OPENID_PERMISSIONS;
+
+public class AccessTokenManager {
+    private static final Logger LOGGER = LoggerFactory.getLogger(AccessTokenManager.class);
+    // We use "aadfeed5" as suffix when client library is ADAL, upgrade to "aadfeed6" for MSAL
+    private static final String REQUEST_ID_SUFFIX = "aadfeed6";
+    private static final String ACCESS_TOKEN = "ACCESS_TOKEN";
+    private static final long ACCESS_TOKEN_MIN_LIVE_TIME = 60 * 1000;
+    private final ServiceEndpoints serviceEndpoints;
+    private final AADAuthenticationProperties aadAuthenticationProperties;
+
+    public AccessTokenManager(ServiceEndpoints serviceEndpoints,
+                              AADAuthenticationProperties aadAuthenticationProperties) {
+        this.serviceEndpoints = serviceEndpoints;
+        this.aadAuthenticationProperties = aadAuthenticationProperties;
+    }
+
+    /**
+     * Acquire AccessToken token for a web-hosted resource with expected permissions. Only work for session supported
+     * web service.
+     *
+     * @param applicationIdUri The Application ID URI of web-hosted resource, e.g., https://graph.microsoft.com for
+     * Microsoft Graph API.
+     * @param permissions The expected permissions of resources.
+     * @return AccessToken
+     * @throws ServiceUnavailableException If fail to acquire the token.
+     */
+    public AccessToken getAccessToken(String applicationIdUri,
```

Comment:
Shall we add a another function of getAccessToken(Set<String> permissions) to cover the case of  user's acquiring openid tokens only?